_drafts/aospobservability.md

```diff
@@ -0,0 +1,102 @@
+
+While there’s a lot of debate about whether logging in production is harmful to device performance or not, it’s undeniable that there’s a lot of value in logs (and a lot of noise). So logs end up being the first tool that developers fall back to when debugging a problem. Except that logs are ephemeral, so inevitably developers will need build a way to cache more logs than just the default of 4MB (depending on the Android version). In my case, we had developed a way to pull the logs off a device if my laptop was connected to the same Wi-Fi network, so early on I would connect to our alpha user’s network to do some debugging.
+
+### Push vs Pull
+
+This might’ve been sufficient for a while (a very short while), but we thought a better way was to have users send us logs when they encounter a problem, so we created a bug report flow that would zip the log cache and then send an email with the logs attached, plus some information about the device and user.
+
+As time went on though, we started running into more and more situations where the cached logs didn’t actually span the time of the incident, so they were basically useless. This is the natural result of a reactive pull-based model, and so we got to brainstorming what other solutions were possible.
+
+We eventually landed on the desire for a push-based model - literally a push messaging system, where we could poke the device remotely to trigger an upload. And we designed an app that could query FileProviders and ContentProviders for any additional data. But we hit a couple of roadblocks: our device was not GMS-compliant so we didn’t have Firebase Cloud Messaging and would have to integrate our own push messaging system, and we needed to enlist the help of a backend team to build the push backend and the push trigger, but also a storage solution, and no team had the budget to help and our teams didn’t have the expertise to contribute. This ends up being a common sticking point with internal tools for mobile teams - Android and AOSP developers have a lot of mobile experience and expertise, but often need help creating a maintainable backend solution that makes sense for the volume and cost of storing and processing the data.
+
+![The logs that are pulled might not contain useful data from around the time of the incident](/img/aosp-observability/logs-buffer.svg)
+
+### Android Bug Reports
+
+Logs are one tool to diagnose an issue, but Android has so much more information if you can access it - namely, Android Bug Reports, which gathers data from all over the device into a single zip file. So any AOSP observability solution that can capture and upload logcat logs automatically, may also want to build a way to trigger and upload Android Bug Reports when needed, too. And so Bug Reports are often the second thing that developers build when they just need more AOSP observability.
```

Comment:
This section makes it sounds like bugreports are the answer (not sure how deep you want to go into the alternatives, but maybe mention e.g. that they contain a complete `dumpsys` dump, which could also be captured individually)